### PR TITLE
chore: Instance Profile V2 migration

### PR DESCRIPTION
**Pull request by edibble21**
Fixes #N/A

**Description**
Migrated Instance Profile Provider V2 Migration and updated tests.

**How was this change tested?**
make presubmit
make test
make ci-test

Does this change impact docs?

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -211,6 +244,19 @@ func CheckEC2Connectivity(ctx context.Context, api ec2iface.EC2API) error {
 	return err
 }
 
+// CheckEC2Connectivity makes a dry-run call to DescribeInstanceTypes.  If it fails, we provide an early indicator that we
+// are having issues connecting to the EC2 API.
+func CheckEC2ConnectivityV2(ctx context.Context, api *ec2V2.Client) error {
```
Comment (engedaam) [inline pkg/operator/operator.go:249]: I don't think we need this update for this PR

**Review by engedaam — COMMENTED** [review]
Nice work 🚀

File: pkg/operator/operator.go [hunk 2]
```diff
@@ -174,11 +179,31 @@ func NewOperator(ctx context.Context, operator *operator.Operator) (context.Cont
 		launchTemplateProvider,
 	)
 
+	//v2
+	cfg, err := configV2.LoadDefaultConfig(ctx,
+		configV2.WithRetryMaxAttempts(5),
```
Comment (engedaam) [inline pkg/operator/operator.go:184]: Why is the retry set to 5? I think we should use the default value for now

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -174,11 +179,31 @@ func NewOperator(ctx context.Context, operator *operator.Operator) (context.Cont
 		launchTemplateProvider,
 	)
 
+	//v2
+	cfg, err := configV2.LoadDefaultConfig(ctx,
```
Comment (engedaam) [inline pkg/operator/operator.go:183]: Let's use `lo.Must` here

File: pkg/operator/operator.go [hunk 4]
```diff
@@ -174,11 +179,31 @@ func NewOperator(ctx context.Context, operator *operator.Operator) (context.Cont
...
+		metaDataClient := imds.NewFromConfig(cfg)
+		region, err := metaDataClient.GetRegion(ctx, nil)
+		if err != nil {
+			log.FromContext(ctx).Error(err, "failed to get region from metadata server")
+			os.Exit(1)
+		}
+		cfg.Region = region.Region
+	}
+
+	instanceProfileProvider := instanceprofile.NewDefaultProvider(cfg.Region, iamV2.NewFromConfig(cfg), cache.New(awscache.InstanceProfileTTL, awscache.DefaultCleanupInterval))
```
Comment (engedaam) [inline pkg/operator/operator.go:201]: Can we leave this one the line it was defined before? Trying to think the best way to structure this files would be

File: pkg/operator/operator.go [hunk 5]
```diff
@@ -130,10 +136,24 @@ func NewOperator(ctx context.Context, operator *operator.Operator) (context.Cont
...
+	//v2
+	cfg := lo.Must(configV2.LoadDefaultConfig(ctx))
+	prometheusv2.WithPrometheusMetrics(cfg, crmetrics.Registry)
+	if cfg.Region == "" {
+		log.FromContext(ctx).V(1).Info("retrieving region from IMDS")
+		metaDataClient := imds.NewFromConfig(cfg)
+		region, err := metaDataClient.GetRegion(ctx, nil)
+		if err != nil {
+			log.FromContext(ctx).Error(err, "failed to get region from metadata server")
+			os.Exit(1)
```
Comment (engedaam) [inline pkg/operator/operator.go:148]: We should do a `lo.Must` instead of an os.Exit(1)

File: pkg/operator/operator.go [hunk 6]
```diff
@@ -130,10 +136,24 @@ func NewOperator(ctx context.Context, operator *operator.Operator) (context.Cont
 		log.FromContext(ctx).WithValues("kube-dns-ip", kubeDNSIP).V(1).Info("discovered kube dns")
 	}
 
+	//v2
+	cfg := lo.Must(configV2.LoadDefaultConfig(ctx))
```
Comment (engedaam) [inline pkg/operator/operator.go:140]: We should define and use the default retry logic https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/retries-timeouts/#standard-retryer

File: pkg/fake/iamapi.go [hunk 7]
```diff
@@ -39,126 +38,188 @@ type IAMAPIBehavior struct {
...
 	TagInstanceProfileBehavior            MockedFunction[iam.TagInstanceProfileInput, iam.TagInstanceProfileOutput]
 	RemoveRoleFromInstanceProfileBehavior MockedFunction[iam.RemoveRoleFromInstanceProfileInput, iam.RemoveRoleFromInstanceProfileOutput]
+	UntagInstanceProfileBehavior          MockedFunction[iam.UntagInstanceProfileInput, iam.UntagInstanceProfileOutput]
 }
 
 type IAMAPI struct {
 	sync.Mutex
 
-	iamiface.IAMAPI
+	IAMClient *iam.Client
```
Comment (engedaam) [inline pkg/fake/iamapi.go:47]: Shouldn't this use `sdk.IAMAPI`?

File: pkg/fake/iamapi.go [hunk 8]
```diff
@@ -39,126 +38,188 @@ type IAMAPIBehavior struct {
...
-		return nil, awserr.New(iam.ErrCodeNoSuchEntityException, fmt.Sprintf("Instance Profile %s cannot be found", aws.StringValue(input.InstanceProfileName)), nil)
+		return nil, &smithy.GenericAPIError{
+			Code: "NoSuchEntityException",
+			Message: fmt.Sprintf("Instance Profile %s cannot be found",
+				aws.ToString(input.InstanceProfileName)),
+		}
+	})
+}
+
+func (s *IAMAPI) UntagInstanceProfile(ctx context.Context, input *iam.UntagInstanceProfileInput, _ ...func(*iam.Options)) (*iam.UntagInstanceProfileOutput, error) {
```
Comment (engedaam) [inline pkg/fake/iamapi.go:208]: Why are we defining `UntagInstanceProfile`? I can't find where we use it?

**Review by engedaam — COMMENTED** [review]
few nits

File: pkg/fake/iamapi.go [hunk 9]
```diff
@@ -39,126 +40,169 @@ type IAMAPIBehavior struct {
 	AddRoleToInstanceProfileBehavior      MockedFunction[iam.AddRoleToInstanceProfileInput, iam.AddRoleToInstanceProfileOutput]
 	TagInstanceProfileBehavior            MockedFunction[iam.TagInstanceProfileInput, iam.TagInstanceProfileOutput]
 	RemoveRoleFromInstanceProfileBehavior MockedFunction[iam.RemoveRoleFromInstanceProfileInput, iam.RemoveRoleFromInstanceProfileOutput]
+	UntagInstanceProfileBehavior          MockedFunction[iam.UntagInstanceProfileInput, iam.UntagInstanceProfileOutput]
```
Comment (engedaam) [inline pkg/fake/iamapi.go:43]: Do we still need this field?

File: pkg/fake/iamapi.go [hunk 10]
```diff
@@ -39,126 +40,169 @@ type IAMAPIBehavior struct {
...
 }
 
 func NewIAMAPI() *IAMAPI {
-	return &IAMAPI{InstanceProfiles: map[string]*iam.InstanceProfile{}}
+	return &IAMAPI{InstanceProfiles: map[string]*iamtypes.InstanceProfile{}}
 }
 
 // Reset must be called between tests otherwise tests will pollute
 // each other.
+func toSliceOfValues[T any](value []T) []T {
```
Comment (engedaam) [inline pkg/fake/iamapi.go:61]: Can't we use `aws.StringSlice`? https://github.com/aws/aws-sdk-go-v2/blob/c5e364e134b6207b81c762ad18f41d09ec531f21/aws/from_ptr.go#L62

File: pkg/fake/iamapi.go [hunk 11]
```diff
@@ -39,126 +40,169 @@ type IAMAPIBehavior struct {
...
 	s.CreateInstanceProfileBehavior.Reset()
 	s.DeleteInstanceProfileBehavior.Reset()
 	s.AddRoleToInstanceProfileBehavior.Reset()
 	s.RemoveRoleFromInstanceProfileBehavior.Reset()
-	s.InstanceProfiles = map[string]*iam.InstanceProfile{}
+	s.InstanceProfiles = map[string]*iamtypes.InstanceProfile{}
 }
 
-func (s *IAMAPI) GetInstanceProfileWithContext(_ context.Context, input *iam.GetInstanceProfileInput, _ ...request.Option) (*iam.GetInstanceProfileOutput, error) {
+func (s *IAMAPI) GetInstanceProfile(ctx context.Context, input *iam.GetInstanceProfileInput, _ ...func(*iam.Options)) (*iam.GetInstanceProfileOutput, error) {
```
Comment (engedaam) [inline pkg/fake/iamapi.go:76]: nit 

It does not seem like we use the context. So we can leave these as they were

```suggestion
func (s *IAMAPI) GetInstanceProfile(_ context.Context, input *iam.GetInstanceProfileInput, _ ...func(*iam.Options)) (*iam.GetInstanceProfileOutput, error) {
```

File: pkg/operator/operator.go [hunk 12]
```diff
@@ -22,6 +22,13 @@ import (
 	"net"
 	"os"
 
+	//v2
```
Comment (engedaam) [inline pkg/operator/operator.go:25]: I think It's easy to understand the deps here. I don't think we need these commnets

**Review by engedaam — COMMENTED** [review]
Few nits

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

File: pkg/fake/iamapi.go (hunk 10 above)
Comment (edibble21) [reply pkg/fake/iamapi.go:61]: StringSlice returns []*string but it needs []string

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

File: pkg/aws/sdk/sdk.go
Comment (rschalo) [inline pkg/aws/sdk/sdk.go:1]: Can we not nest this? The file path stutters.

File: pkg/operator/operator.go [hunk 13]
```diff
@@ -130,10 +133,19 @@ func NewOperator(ctx context.Context, operator *operator.Operator) (context.Cont
 		log.FromContext(ctx).WithValues("kube-dns-ip", kubeDNSIP).V(1).Info("discovered kube dns")
 	}
 
+	//v2
+	cfg := lo.Must(configV2.LoadDefaultConfig(ctx, configV2.WithRetryMaxAttempts(3)))
+	if cfg.Region == "" {
+		log.FromContext(ctx).V(1).Info("retrieving region from IMDS")
+		metaDataClient := imds.NewFromConfig(cfg)
+		region := lo.Must(metaDataClient.GetRegion(ctx, nil))
+		cfg.Region = region.Region
```
Comment (engedaam) [inline pkg/operator/operator.go:142]: Can we inline these two lines?

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

File: pkg/operator/operator.go [hunk 14]
```diff
@@ -130,10 +133,19 @@ func NewOperator(ctx context.Context, operator *operator.Operator) (context.Cont
 		log.FromContext(ctx).WithValues("kube-dns-ip", kubeDNSIP).V(1).Info("discovered kube dns")
 	}
 
+	//v2
+	cfg := lo.Must(configV2.LoadDefaultConfig(ctx, configV2.WithRetryMaxAttempts(3)))
```
Comment (engedaam) [inline pkg/operator/operator.go:137]: Let's add a comment on adding prometheus metrics here

File: pkg/operator/operator.go [hunk 15]
```diff
@@ -130,10 +133,19 @@ func NewOperator(ctx context.Context, operator *operator.Operator) (context.Cont
 		log.FromContext(ctx).WithValues("kube-dns-ip", kubeDNSIP).V(1).Info("discovered kube dns")
 	}
 
+	//v2
+	//Once everything is migrated we will need to update prometheus metrics to v2
+	cfg := lo.Must(configV2.LoadDefaultConfig(ctx, configV2.WithRetryMaxAttempts(3)))
+	if cfg.Region == "" {
+		log.FromContext(ctx).V(1).Info("retrieving region from IMDS")
+		metaDataClient := imds.NewFromConfig(cfg)
+		cfg.Region = lo.Must(metaDataClient.GetRegion(ctx, nil)).Region
```
Comment (engedaam) [inline pkg/operator/operator.go:142]: As a follow-up, we should add in an error message here https://github.com/aws/karpenter-provider-aws/pull/7097/files#diff-99c0290799daf9abc6240df64063e20bfaf67b371577b67ac7eec6f4725622ffR111

**Review by engedaam — APPROVED** [review]
LGTM 🚀

**Review by engedaam — APPROVED** [review]
LGTM 🚀

**Pull request merged**